scantool/diag_l0_elm.c

```diff
@@ -743,13 +752,77 @@ elm_initbus(struct diag_l0_device *dl0d, struct diag_l1_initbus_args *in)
 					break;
 			}
 
+			//set init address
+			if ((dev->elmflags & ELM_323_BASIC) && (in->addr != 0x33)) {
+				fprintf(stderr, FLFMT "elm_initbus: ELM323 doesn't support target address %02X", FL, in->addr);
+				return diag_iseterr(DIAG_ERR_GENERAL);
+			} else {
+				uint8_t iia[15];
+				sprintf((char *) iia, "ATIIA %02X\x0D", in->addr);
+				rv=elm_sendcmd(dl0d, iia, 9, 500, NULL);
+				if (rv < 0 && in->addr == 0x33) {
+					fprintf(stderr, FLFMT "elm_initbus: ATIIA 33 failed, continuing anyway\n", FL);
+					rv = 0;
+				} else if (rv < 0) {
+					fprintf(stderr, FLFMT "elm_initbus: ATIIA %02X failed\n", FL, in->addr);
+					return diag_iseterr(DIAG_ERR_GENERAL);
+				}
+			}
+
+			//accept nonstandard keybytes during init
+			if (dev->elmflags & ELM_327_BASIC) {
+				buf=(uint8_t *) "ATKW0\x0D";
+				rv=elm_sendcmd(dl0d, buf, 6, 500, NULL);
+				if (rv < 0)
+					fprintf(stderr, FLFMT "elm_initbus: ATKW0 failed, continuing anyway\n", FL);
+			}
+
 			//explicit init is not supported by clones
 			if ((dev->elmflags & ELM_32x_CLONE)==0) {
 				rv = elm_slowinit(dl0d);
 				if (!rv)
 					dev->elmflags |= ELM_INITDONE;
 			}
 			timeout=4200;	//slow init is slow !
+
+			//query elm for keybytes
+			dev->kb1 = 0; dev->kb2 = 0;
+			if (dev->elmflags & ELM_INITDONE) {
+				uint8_t rxbuf[ELM_BUFSIZE];
+				unsigned int kb1, kb2;
+
+				buf=(uint8_t *) "ATKW\x0D";
+				rv=elm_sendcmd(dl0d, buf, 5, 500, rxbuf);
+				if (rv < 0) {
+					fprintf(stderr, FLFMT "elm_initbus: ATKW failed, continuing anyway\n", FL);
+					// TODO: for clones without ATKW, can
+					// fall back to ATBD. If ATBD response
+					// starts with 06 addr 55 xx yy aa bb
+					// and aa=yy^0xff and bb=addr^0xff then
+					// xx and yy should be key bytes
+					rv = 0;
+				} else if (sscanf((char *)rxbuf, "1:%02X 2:%02X", &kb1, &kb2) == 2) {
+					dev->kb1 = kb1;
+					dev->kb2 = kb2;
+					in->kb1 = kb1;
+					in->kb2 = kb2;
+				} else {
+					fprintf(stderr, FLFMT "elm_initbus: ATKW failed, continuing anyway\n", FL);
+					rv = 0;
+				}
+			}
+
+			//for KWP6227, set appropriate wakeup message
```

Comment:
hmm, I know this is extra work, but I'd really prefer having this as another IOCTL. It could be quite simple and only implemented on ELM for now, and would move the ugliness of hardcoded constants out of L0 code and into the L2 code.